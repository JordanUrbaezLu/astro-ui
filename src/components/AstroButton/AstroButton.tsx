import * as React from "react";
import styles from "./AstroButton.module.scss";
import classNames from "classnames";

export type AstroButtonVariant = "primary" | "secondary";
export type AstroButtonSize = "small" | "medium" | "large";
export type AstroButtonType = "button" | "submit" | "reset";

export interface AstroButtonProps {
  /**
   * The content for the AstroButton
   */
  children: React.ReactNode;
  /**
   * If the AstroButton is disabled
   */
  disabled?: boolean;
  /**
   * The leading icon for the AstroButton
   */
  leadingIcon?: React.ReactNode;
  /**
   * The callback fired when the AstroButton is clicked
   */
  onClick?: () => void;
  /**
   * The size of the AstroButton
   *
   * @default "medium"
   */
  size?: AstroButtonSize;
  /**
   * The trailing icon for the AstroButton
   */
  trailingIcon?: React.ReactNode;
  /**
   * The type of the AstroButton
   */
  type?: AstroButtonType;
  /**
   * The variant of the AstroButton
   *
   * @default "primary"
   */
  variant?: AstroButtonVariant;
}

const AstroButton = React.forwardRef<HTMLButtonElement, AstroButtonProps>(
  (props, ref) => {
    const {
      children,
      disabled,
      leadingIcon,
      onClick,
      size = "medium",
      trailingIcon,
      type,
      variant = "primary",
      ...rest
    } = props;

    const classes = classNames(
      styles.astroButton,
      size === "small"
        ? styles.small
        : size === "large"
        ? styles.large
        : styles.medium,
      variant === "secondary" ? styles.secondary : styles.primary,
      disabled && styles.disabled
    );

    const content = (
      <>
        {leadingIcon && (
          <span className={styles.leadingIcon}>{leadingIcon}</span>
        )}
        {children}
        {trailingIcon && (
          <span className={styles.trailingIcon}>{trailingIcon}</span>
        )}
      </>
    );

    return (
      <button
        className={classes}
        onClick={onClick}
        disabled={disabled}
        ref={ref}
        type={
          type === "reset" ? "reset" : type === "submit" ? "submit" : "button"
        }
        {...rest}
      >
        {content}
      </button>
    );
  }
);

export default AstroButton;
