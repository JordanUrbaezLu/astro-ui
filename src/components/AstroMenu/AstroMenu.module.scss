.container {
  display: flex;
  flex-direction: column;
  position: absolute;
}

.astroMenuContainer {
  display: flex;
  flex-direction: column;
  position: relative;
  background-color: rgb(50, 50, 50);
  border-radius: 4px;
  padding: 6px 0px;
  z-index: 999;
  box-shadow: rgba(0, 0, 0, 0.35) 0px 5px 15px;
  transform: translateY(2px);
}

.defaultTriggerContainer {
  cursor: pointer;
  color: white;
  font-size: 20px;
  background-color: rgb(50, 50, 50);
  border-radius: 999px;
  padding: 8px;
  &:hover,
  &:focus {
    background-color: rgb(80, 80, 80);
  }
}

.enter {
  opacity: 0;
  transform: translateY(-8px);
}

.enterActive {
  transform: translateY(2px);
  opacity: 1;
  transition: all 150ms ease-in-out;
}

.exit {
  transform: translateY(2px);
  opacity: 1;
}

.exitActive {
  transform: translateY(-8px);
  opacity: 0;
  transition: all 150ms ease-in-out;
}
