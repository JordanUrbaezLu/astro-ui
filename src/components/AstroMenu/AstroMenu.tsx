import * as React from "react";
import { CSSTransition } from "react-transition-group";
import { AiOutlineEllipsis } from "react-icons/ai";
import styles from "./AstroMenu.module.scss";

export interface AstroMenuProps {
  /**
   * The content for the AstroMenu
   */
  children: React.ReactNode;
  /**
   * The callback fired when requested to change state
   */
  setValue?: React.Dispatch<React.SetStateAction<string>>;
  /**
   * The trigger for the AstroMenu
   */
  trigger?: React.ReactElement;
}

const AstroMenu: React.FC<AstroMenuProps> = (props) => {
  const { children, setValue, trigger } = props;

  const [isOpen, setIsOpen] = React.useState<boolean>(false);

  const [focusedMenuItem, setFocusedMenuItem] = React.useState<number>(0);

  const menuRef = React.useRef<HTMLDivElement>(null);

  const lastIndex = React.Children.count(children) - 1;

  const focusNextItem = () => {
    setFocusedMenuItem((prevIndex) =>
      prevIndex + 1 > lastIndex ? 0 : prevIndex + 1
    );
  };

  const focusPreviousItem = () => {
    setFocusedMenuItem((prevIndex) =>
      prevIndex - 1 < 0 ? lastIndex : prevIndex - 1
    );
  };

  React.useEffect(() => {
    if (isOpen) {
      const child = menuRef.current?.children[focusedMenuItem] as HTMLElement;
      child?.focus();
    }
  }, [isOpen, focusedMenuItem]);

  return (
    <div className={styles.container}>
      {trigger ? (
        React.cloneElement(trigger, {
          onClick: () => setIsOpen(!isOpen),
        })
      ) : (
        <AiOutlineEllipsis
          className={styles.defaultTriggerContainer}
          onClick={() => setIsOpen(!isOpen)}
          onKeyDown={(event) => {
            if (event.code === "Enter") {
              setIsOpen(!isOpen);
            }
          }}
          role="button"
          tabIndex={0}
        />
      )}
      <CSSTransition
        in={isOpen}
        timeout={150}
        mountOnEnter
        unmountOnExit
        classNames={{
          enter: styles.enter,
          enterActive: styles.enterActive,
          exit: styles.exit,
          exitActive: styles.exitActive,
        }}
      >
          <div className={styles.astroMenuContainer} ref={menuRef} role="menu">
            {React.Children.map(children, (child, index) => {
              if (
                React.isValidElement<React.HTMLAttributes<HTMLElement>>(child)
              ) {
                return React.cloneElement(child, {
                  key: index,
                  onClick: () => {
                    if (setValue) {
                      setValue(child.props.children as string);
                    }
                    setIsOpen(false);
                  },
                  onKeyDown: (event: any) => {
                    if (event.code === "Enter" || event.code === "Escape") {
                      setIsOpen(false);
                    }
                    if (event.code === "ArrowUp") {
                      focusPreviousItem();
                    }
                    if (event.code === "ArrowDown") {
                      focusNextItem();
                    }
                  },
                  role: "menuitem",
                  tabIndex: 0,
                });
              } else {
                return null;
              }
            })}
          </div>
      </CSSTransition>
    </div>
  );
};

export default AstroMenu;
