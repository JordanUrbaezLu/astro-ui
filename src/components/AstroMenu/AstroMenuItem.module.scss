.astroMenuItemContainer {
  color: white;
  cursor: pointer;
  font-family: Verdana, Geneva, Tahoma, sans-serif;
  padding: 0px 8px;
  user-select: none;
  &:hover,
  &:focus {
    background-color: rgb(80, 80, 80);
  }
}
