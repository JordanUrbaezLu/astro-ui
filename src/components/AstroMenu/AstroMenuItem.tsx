import * as React from "react";
import styles from "./AstroMenuItem.module.scss";

export interface AstroMenuItemProps {
  /**
   * The content for the AstroMenuItem
   */
  children: string;
}

const AstroMenuItem: React.FC<AstroMenuItemProps> = (props) => {
  const { children, ...rest } = props;

  return (
    <div className={styles.astroMenuItemContainer} {...rest}>
      {children}
    </div>
  );
};

export default AstroMenuItem;
