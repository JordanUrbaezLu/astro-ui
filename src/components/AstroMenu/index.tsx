export * from "./AstroMenu";
export * from "./AstroMenuItem";
