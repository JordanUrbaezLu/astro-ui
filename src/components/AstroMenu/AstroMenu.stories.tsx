import { Story } from "@storybook/react";
import * as React from "react";
import AstroMenu, { AstroMenuProps } from "./AstroMenu";
import AstroMenuItem from "./AstroMenuItem";
import AstroButton from "../AstroButton/AstroButton";

export default {
  component: AstroMenu,
  title: "Components/AstroMenu",
};

const Template: Story<
  Omit<
    AstroMenuProps,
    "children" | "isOpen" | "onClose" | "onOpen" | "trigger"
  >
> = (args) => {
  const [value, setValue] = React.useState<string | undefined>(undefined);
  const [isOpen, setIsOpen] = React.useState<boolean>(true);

  return (
    <>
      <div
        style={{ fontFamily: "Verdana", paddingBottom: "4px", width: "200px" }}
      >
        Value: {value}
      </div>
      <AstroMenu
        {...args}
        isOpen={isOpen}
        onOpen={() => setIsOpen(true)}
        onClose={(itemValue) => {
          setIsOpen(false);
          if (itemValue !== undefined) {
            setValue(itemValue);
          }
        }}
        trigger={<AstroButton>Trigger</AstroButton>}
      >
        <AstroMenuItem>Menu Item One</AstroMenuItem>
        <AstroMenuItem>Menu Item Two</AstroMenuItem>
        <AstroMenuItem>Menu Item Three</AstroMenuItem>
        <AstroMenuItem>Menu Item Four</AstroMenuItem>
      </AstroMenu>
    </>
  );
};

export const Default = Template.bind({});

export const AlignmentBottomLeft = Template.bind({});
AlignmentBottomLeft.args = {
  alignment: "bottomLeft",
};

export const AlignmentBottomRight = Template.bind({});
AlignmentBottomRight.args = {
  alignment: "bottomRight",
};

export const AlignmentBottomCenter = Template.bind({});
AlignmentBottomCenter.args = {
  alignment: "bottomCenter",
};

export const AlignmentTopLeft = Template.bind({});
AlignmentTopLeft.args = {
  alignment: "topLeft",
};

export const AlignmentTopRight = Template.bind({});
AlignmentTopRight.args = {
  alignment: "topRight",
};

export const AlignmentTopCenter = Template.bind({});
AlignmentTopCenter.args = {
  alignment: "topCenter",
};
