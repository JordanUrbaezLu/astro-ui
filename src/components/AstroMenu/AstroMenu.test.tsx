import { fireEvent, render, screen } from "@testing-library/react";
import AstroMenu from "./AstroMenu";
import AstroMenuItem from "./AstroMenuItem";
import { axe } from "jest-axe";
import AstroButton from "../AstroButton/AstroButton";

describe("AstroMenu", () => {
  test("Should render component correctly.", () => {
    const { container } = render(
      <AstroMenu
        isOpen
        onOpen={jest.fn}
        onClose={jest.fn()}
        trigger={<AstroButton>Trigger</AstroButton>}
      >
        <AstroMenuItem>One</AstroMenuItem>
      </AstroMenu>
    );
    expect(container.firstChild).toMatchSnapshot();
  });
});

describe("Accessibility", () => {
  test("Should have no accessibility violations.", async () => {
    const { container } = render(
      <AstroMenu
        isOpen
        onOpen={jest.fn}
        onClose={jest.fn()}
        trigger={<AstroButton>Trigger</AstroButton>}
      >
        <AstroMenuItem>One</AstroMenuItem>
      </AstroMenu>
    );
    const results = await axe(container);
    expect(results).toHaveNoViolations();
  });
});

describe("Props", () => {
  test("Should render children correctly.", () => {
    render(
      <AstroMenu
        isOpen
        onOpen={jest.fn}
        onClose={jest.fn()}
        trigger={<AstroButton>Trigger</AstroButton>}
      >
        <AstroMenuItem>One</AstroMenuItem>
      </AstroMenu>
    );
    fireEvent.click(screen.getByRole("button"));
    expect(screen.getByRole("menuitem")).toHaveTextContent("One");
  });

  test("Should render trigger correctly.", () => {
    render(
      <AstroMenu
        isOpen
        onOpen={jest.fn}
        onClose={jest.fn()}
        trigger={<AstroButton>Trigger</AstroButton>}
      >
        <AstroMenuItem>One</AstroMenuItem>
      </AstroMenu>
    );
    expect(screen.getAllByRole("button")[0]).toHaveTextContent("Trigger");
  });
});
