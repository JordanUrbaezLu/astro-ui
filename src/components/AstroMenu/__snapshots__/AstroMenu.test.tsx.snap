// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AstroMenu Should render component correctly. 1`] = `
<div
  class="container"
>
  <button
    class="astroButton medium primary"
    role="button"
    tabindex="0"
    type="button"
  >
    Trigger
  </button>
  <div
    class="astroMenuLayoutContainer"
    style="transform: translate3d(0px, 0px, 0);"
  >
    <div
      class="astroMenuContainer"
      role="menu"
    >
      <div
        class="astroMenuItemContainer"
        role="menuitem"
        tabindex="0"
      >
        One
      </div>
    </div>
  </div>
</div>
`;
